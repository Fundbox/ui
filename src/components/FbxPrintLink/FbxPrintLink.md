# FbxPrintLink

## Basic usage

- Use this component to open printable content in a new window
- Generally used for printing agreements

```html
<fbx-print-link
  print-title="A title"
  print-content="Some content"
/>
```
