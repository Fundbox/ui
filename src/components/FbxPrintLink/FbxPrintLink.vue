<template>
  <div class="fbx-print-link">
    <div v-if="withIcon" class="fbx-icon-printer"></div>
    <a href="#" @click="onPrintButtonClick">Print this agreement</a>
  </div>
</template>

<script>
import escape from 'lodash/escape'
import xss from 'xss'

export default {
  name: 'FbxPrintLink',
  props: {
    printTitle: {
      type: String,
      required: true,
    },
    printHtmlContent: {
      type: String,
      required: true,
    },
    withIcon: {
      type: Boolean,
      default: true,
    },
  },
  methods: {
    onPrintButtonClick() {
      const htmlToPrint = `
        <!DOCTYPE html>
        <html lang="en">
          <head>
            <meta charset="UTF-8">
            <title>${escape(this.printTitle)}</title>
          </head>
          <body>
            <h1>${escape(this.printTitle)}</h1>
            <div>${xss(this.printHtmlContent)}</div>
          </body>
        </html>
      `
      const newWindow = window.open()
      newWindow.document.documentElement.innerHTML = htmlToPrint
      newWindow.focus()
      newWindow.print()
    },
  },
}
</script>

 <style lang="scss" scoped>
@import "../../styles/utils/color-palette.scss";

.fbx-print-link {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
  color: $dark-green;
}

.fbx-icon-printer {
  margin-right: 5px;
}
</style>
