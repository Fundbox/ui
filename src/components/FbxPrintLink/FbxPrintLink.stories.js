import { storiesOf } from '@storybook/vue'
import { withInfo } from 'storybook-addon-vue-info'

import FbxPrintLink from './FbxPrintLink.vue'
import summary from './FbxPrintLink.md'

const stories = storiesOf('Components/PrintLink', module)

stories.add('default', withInfo({ summary })(() => ({
  components: { FbxPrintLink },
  data() {
    return {
      printTitle: 'Ein Cupcakelied',
      printHtmlContent: `
        Cupcake ipsum dolor sit amet cake jelly-o.
        Topping biscuit donut biscuit jelly beans candy danish.
        Cheesecake macaroon pastry sweet roll bear claw powder dragée macaroon.
        Icing macaroon jelly beans sweet roll.
        Cake chocolate cupcake cotton candy jujubes tootsie roll croissant toffee.
        Marzipan sesame snaps lollipop dessert cupcake fruitcake tiramisu icing marshmallow.
        Fruitcake muffin danish fruitcake. Cotton candy toffee pastry sweet sweet sweet chocolate cake halvah.
        Pie soufflé candy canes tootsie roll candy chocolate bar cookie. Chocolate bar icing cupcake pastry tart pudding.
      `,
    }
  },
  template: `
    <div>
      <div style="margin-bottom: 25px;">
        <fbx-print-link
          :print-title="printTitle"
          :print-html-content="printHtmlContent"
        />
      </div>

      <p style="margin-bottom: 10px;">Without print icon:</p>

      <fbx-print-link
        :print-title="printTitle"
        :print-html-content="printHtmlContent"
        :with-icon="false"
      />
    </div>
  `,
})))
