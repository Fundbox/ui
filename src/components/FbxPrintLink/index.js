import FbxPrintLink from './FbxPrintLink.vue'
import { registerComponents, vueUse } from '../../utils/plugins'

const components = {
  FbxPrintLink,
}

const VuePlugin = {
  install(Vue) {
    registerComponents(Vue, components)
  }
}

vueUse(VuePlugin)

export default VuePlugin
export { FbxPrintLink }
