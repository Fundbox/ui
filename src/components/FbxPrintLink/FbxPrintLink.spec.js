import { shallowMount } from '@vue/test-utils'
import FbxPrintLink from './FbxPrintLink.vue'

describe('Components/FbxPrintLink', () => {
  const printTitle = 'A title'
  const printHtmlContent = 'Some content'

  describe('snapshots', () => {
    it('renders the default correctly', () => {
      const wrapper = shallowMount(FbxPrintLink, {
        propsData: {
          printTitle,
          printHtmlContent,
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders without a print icon', () => {
      const wrapper = shallowMount(FbxPrintLink, {
        propsData: {
          printTitle,
          printHtmlContent,
          showPrintIcon: false,
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('onPrintButtonClick', () => {
    it('prints the new document', () => {
      const focus = jest.fn()
      const print = jest.fn()

      window.open = jest.fn(() => {
        return {
          document: {
            documentElement: {
              innerHTML: '',
            },
          },
          focus,
          print,
        }
      })

      const wrapper = shallowMount(FbxPrintLink, {
        propsData: {
          printTitle,
          printHtmlContent,
        },
      })

      wrapper.vm.onPrintButtonClick()

      const finalPrintContent = window.open.mock.results[0].value.document.documentElement.innerHTML
      expect(finalPrintContent).toContain('DOCTYPE')
      expect(finalPrintContent).toContain(printTitle)
      expect(finalPrintContent).toContain(printHtmlContent)
      expect(window.open).toHaveBeenCalledTimes(1)
      expect(focus).toHaveBeenCalledTimes(1)
      expect(print).toHaveBeenCalledTimes(1)
    })
  })
})
