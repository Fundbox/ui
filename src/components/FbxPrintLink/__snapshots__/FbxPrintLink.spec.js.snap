// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/FbxPrintLink snapshots renders the default correctly 1`] = `
<div class="fbx-print-link">
  <div class="fbx-icon-printer"></div> <a href="#">Print this agreement</a>
</div>
`;

exports[`Components/FbxPrintLink snapshots renders without a print icon 1`] = `
<div class="fbx-print-link">
  <!----> <a href="#">Print this agreement</a></div>
`;
